packages/peregrine/lib/talons/CreateAccount/useCreateAccount.js

```diff
@@ -77,32 +92,31 @@ export const useCreateAccount = props => {
                         is_subscribed: !!formValues.subscribe
                     }
                 });
-
-                // Sign in and save the token
-                const response = await signIn({
+                const signInResponse = await signIn({
                     variables: {
                         email: formValues.customer.email,
                         password: formValues.password
                     }
                 });
-
-                const token =
-                    response && response.data.generateCustomerToken.token;
-
+                const token = signInResponse.data.generateCustomerToken.token;
                 await setToken(token);
-                await getUserDetails({ fetchUserDetails });
-
-                // Then remove the old guest cart and get the cart id from gql.
-                // TODO: This logic may be replacable with mergeCart in 2.3.4
-                await removeCart();
-
-                await clearCartDataFromCache(apolloClient);
-                await clearCustomerDataFromCache(apolloClient);
 
+                // Create and get the customer's cart id.
                 await createCart({
                     fetchCartId
                 });
+                const destinationCartId = await retrieveCartId();
```

Comment:
This all works but I have concerns about directly fetching the cart id from local storage. It's a little magical in that it relies on the logic internal to `createCart` that actually calls the invoker and sets the response value in local storage and in redux. We _must_ make this call right now since the entire application relies on redux having a valid `cartId` at all times.

Ideally we could also return it from the action or maybe get `cartId` back from a query instead of reading from local storage. As we've found though, awaiting query invokers is not what ApolloGQL intended.

Reply:
I've been thinking a lot about getting away from this pattern of chaining, and I want your thoughts if this might clean this logic up, and get rid of the need to inject this into the current logic. This code could even live at a single place closer to the app root, no need to duplicate. Say it with me, react to state!

This is a bit mocked up, so deps are missing, sorry.
```js
const inititalCartId = useRef();

useEffect(() => {
    if (cartId && !inititalCartId.current) {
        inititalCartId.current = cartId;
    }
});

useEffect(() => {
    if (isSignedIn && cartId !== inititalCartId.current) {
        // merge those damn carts
        inititalCartId.current = cartId;
    }
});
```

There might be an edge here where an authenticated user just checked out, so you wouldn't want to attempt a merge, but I'm sure we could handle that. Consider this uber optional if you're looking for an alternative to fetching direct from localStorage.

Reply:
>There might be an edge here where an authenticated user just checked out, so you wouldn't want to attempt a merge, but I'm sure we could handle that

I also think that we're going to have a tough time doing this with the current retry logic in redux. I'd rather revisit this later once we've deprecated the redux cart slice.

> This code could even live at a single place closer to the app root, no need to duplicate.

This. I would love to have the cart context own the code for handling cart creation and merging.